### DISCUSSION TITLE
Detecting method calls where receiver is of given type

### DISCUSSION THREAD

**Discussion by knutwannheden** [TNG/ArchUnit#1122] [Q&A] [answered]
I would like to write a rule where I want to check the type of the receiver of the call. Specifically, I want to check that there is no code which calls `Object#equals(Object)` on any instance of `Foo`. How do I do this? Please note that my class `Foo` does **not** override `equals()` itself.

Of course this rule can only be expected to catch cases where the type of the receiver can statically be determined to be of type `Foo`. So if I have a variable like `Object foo = new Foo();` and then at some point call `foo.equals(whatever)` then the rule won't be able to catch that, but that is fine for my use case.

**Comment by hankem** [comment, marked as answer]
I fear that the byte code might not contain the information that `Object#equals(Object)` was called on a `Foo`:
In the example
```java
class Foo {
   boolean eq() {
     Foo foo = new Foo();
     return foo.equals(new Object());
   }
}
```
the method `eq` is compiled to the byte code (cf. `javap -c -p -v Foo.class`):
```
         0: new           #7                  // class Foo
         3: dup
         4: invokespecial #9                  // Method "<init>":()V
         7: astore_1
         8: aload_1
         9: new           #2                  // class java/lang/Object
        12: dup
        13: invokespecial #1                  // Method java/lang/Object."<init>":()V
        16: invokevirtual #10                 // Method java/lang/Object.equals:(Ljava/lang/Object;)Z
        19: ireturn
```

**Reply by knutwannheden** [reply]
I see. I forgot that ArchUnit is bytecode-based and not AST-based. Thanks for clarifying!
